File: src/scatter_lines_rgbwt.cpp
/*
 * This file is part of scattermore.
 *
 * Copyright (C) 2022 Mirek Kratochvil <exa.exa@gmail.com>
 *               2022-2023 Tereza Kulichova <kulichova.t@gmail.com>
 *
 * scattermore is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * scattermore is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * scattermore. If not, see <https://www.gnu.org/licenses/>.
 */

#include "scatters_lines.h"
#include "scatters_lines_impl.h"

#include <cstdlib>

// draw given lines
void
scatter_lines_rgbwt(const float *xy,
                    const unsigned *dim,
                    const float *xlim,
                    const float *ylim,
                    const float *RGBA,
                    const int *skip,
                    float *RGBWT)
{
  const size_t size_out_x = dim[0];
  const size_t size_out_y = dim[1];
  const size_t size_data = dim[2];
  const size_t size_out = size_out_y * size_out_x;

  const size_t offset_R = size_out * 0;
  const size_t offset_G = size_out * 1;
  const size_t offset_B = size_out * 2;
  const size_t offset_W = size_out * 3;
  const size_t offset_T = size_out * 4;

  const int skip_start_pixel = skip[0];
  const int skip_end_pixel = skip[1];

  const float x_begin = xlim[0];
  const float x_end = xlim[1];
  const float x_bin = (size_out_x - 1) / (x_end - x_begin);

  const float y_begin = ylim[1];
  const float y_end = ylim[0];
  const float y_bin = (size_out_y - 1) / (y_end - y_begin);

  float R = RGBA[0];
  float G = RGBA[1];
  float B = RGBA[2];
  float A = RGBA[3];

  // lambda expression used for individual pixel when plotting a line
  auto pixel_function_rgbwt = [R,
                               G,
                               B,
                               A,
                               size_out_x,
                               size_out_y,
                               offset_R,
                               offset_G,
                               offset_B,
                               offset_W,
                               offset_T,
                               RGBWT](size_t x, size_t y) {
    if (x < size_out_x && y < size_out_y) {
      size_t offset = x * size_out_y + y;
      RGBWT[offset + offset_R] += R * A;
      RGBWT[offset + offset_G] += G * A;
      RGBWT[offset + offset_B] += B * A;
      RGBWT[offset + offset_W] += A;
      RGBWT[offset + offset_T] *= 1 - A;
    }
  };

  for (size_t i = 0; i < size_data; ++i) {
    int x0 = (xy[0 * size_data + i] - x_begin) * x_bin;
    int y0 = (xy[1 * size_data + i] - y_begin) * y_bin;
    int x1 = (xy[2 * size_data + i] - x_begin) * x_bin;
    int y1 = (xy[3 * size_data + i] - y_begin) * y_bin;

    plot_line(
      x0, y0, x1, y1, skip_start_pixel, skip_end_pixel, pixel_function_rgbwt);
  }
}

File: R/scatter_lines_rgbwt.R
# This file is part of scattermore.
#
# Copyright (C) 2022 Mirek Kratochvil <exa.exa@gmail.com>
#               2022-2023 Tereza Kulichova <kulichova.t@gmail.com>
#
# scattermore is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# scattermore is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with scattermore. If not, see <https://www.gnu.org/licenses/>.

#' scatter_lines_rgbwt
#'
#' Render lines into a RGBWT bitmap.
#'
#' @param xy 4-column matrix with point coordinates.
#'           Each row contains X and Y coordinates of line start and X and Y coordinates of line end, in this order.
#'
#' @param xlim,ylim 2-element vector of rendered area limits (position of the first pixel on the
#'                   left/top, and the last pixel on the right/bottom).
#'                   You can flip the image coordinate system by flipping the `*lim` vectors.
#'
#' @param RGBA Vector of 4 elements with integral RGBA color for the lines, defaults to `c(0,0,0,255)`.
#'
#' @param skip_start_pixel TRUE if the start pixel of the lines should be omitted, defaults to `FALSE`.
#'
#' @param skip_end_pixel TRUE if the end pixel of a line should be omitted, defaults to `TRUE`.
#'                       (When plotting long ribbons of connected lines, this prevents counting the connecting pixels twice.)
#'
#' @return Lines plotted in RGBWT bitmap.
#'
#' @export
#' @useDynLib scattermore, .registration=TRUE

scatter_lines_rgbwt <- function(xy,
                                xlim = c(min(xy[,c(1,3)]), max(xy[,c(1,3)])),
                                ylim = c(min(xy[,c(2,4)]), max(xy[,c(2,4)])),
                                out_size = c(512L, 512L),
                                RGBA = c(0, 0, 0, 255),
                                skip_start_pixel = FALSE,
                                skip_end_pixel = TRUE) {
  if (!is.numeric(xlim) || length(xlim) != 2) stop("invalid xlim")
  if (!is.numeric(ylim) || length(ylim) != 2) stop("invalid ylim")
  if (!is.numeric(out_size) || length(out_size) != 2) stop("invalid out_size")

  n <- if ((is.matrix(xy) || is.array(xy)) && dim(xy)[2] == 4) dim(xy)[1]
       else stop("invalid line coordinates in xy (expected 4-column matrix)")

  if (!is.numeric(RGBA) || length(RGBA) != 4) stop("invalid RGBA")

  size_x <- as.integer(out_size[1])
  size_y <- as.integer(out_size[2])

  RGBWT <- array(0, c(size_y, size_x, 5))
  RGBWT[, , 5] <- 1 # initialize the transparency (multiplicative)

  result <- .C("scatter_lines_rgbwt",
    xy = as.single(xy),
    dimen = as.integer(c(size_x, size_y, n)),
    xlim = as.single(xlim),
    ylim = as.single(ylim),
    RGBA = as.single(RGBA / 255),
    skip_pixel = as.integer(c(skip_start_pixel, skip_end_pixel)),
    fRGBWT = as.single(RGBWT)
  )

  return(array(result$fRGBWT, c(size_y, size_x, 5)))
}
